xdsl/ir.py

```diff
@@ -659,6 +659,30 @@ def add_region(self, region: Region) -> None:
         self.regions.append(region)
         region.parent = self
 
+    def get_region_index(self, region: Region) -> int:
+        """Get the region position in the operation."""
+        if region.parent is not self:
+            raise Exception("Region is not inside the operation.")
+        for idx, curr_region in enumerate(self.regions):
+            if curr_region is region:
+                return idx
+        assert False, "Unexpected xdsl error"
+
+    def detach_region(self, region: int | Region) -> Region:
+        """
+        Detach a region from the operation.
+        Returns the detached region.
+        """
+        if isinstance(region, Region):
+            region_idx = self.get_region_index(region)
+        else:
+            region_idx = region
+            region = self.regions[region_idx]
+        region.parent = None
+        self.regions = self.regions[:region_idx] + self.regions[region_idx +
+                                                                1:]
```

Comment:
I am not quite sure this should work like this. If we detach a region and move the other regions back, won't this potentially break everything in the case where you don't have variadic regions, but rather just 3 regions?

Say you have an operation that models a threeway if with an operand deciding which region to take. Say that the value of the operand maps the following:
- negative --> first region
- 0 --> second region
- positive --> third region

If you then detach like this, you would reorder the regions.

I would argue that we should (a) enforce that you are removing a variadic region def or (b) replace the region by an empty one if it does not fall in a variadic region.

Reply:
Indeed, this can get very confusing. (Especially the by-value case: which index got removed? How do I put the meaningful new one in the correct index?)
I would go for something like (b): just implement `replace_region` with an optional new `Region` argument, defaulting to replacing by an empty one?

Reply:
I think there should be a case distinction on whether or not the region that is removed is part of a variadic region. If not, replacing by empty (or an optional argument) would probably make most sense.

Reply:
I believe that this is the interface we want to expose here.
These operations are used for fine-grained operations in xDSL, that we can use with PDL or IRDL for instance.
If we want to remove a region from a variadic, we should do that in `IRDLOperation` instead, or with a different interface.
At least that's my feeling.

Reply:
> If we want to remove a region from a variadic, we should do that in IRDLOperation instead

I agree

Reply:
So you are essentially saying that this is an operation on the data structure, so it should be agnostic of all the IRDL stuff and only know that it has a list of regions, one of which should be removed.

I agree with this take, actually it's probably a better take than my initial one. However, if we decided to go this route, I don't think we should depend on IRDL in the testing part. IMO, we should change the test s.t. it can work without IRDL. I know, this is not something we have done a lot of so far (constructing ops without IRDL at all) and maybe this is not even well supported in the current repo.

Therefore, I think we could merge this as is currently, **but** we need this to be documented in the docstring of the method and we need to open an issue on the xDSL repo detailing this issue/discussion.

Do you guys think this is a sensible way of doing this, @math-fehr @meshtag @PapyChacal?

Reply:
> we need this to be documented in the docstring of the method

Shall I add the following ToDo [here](https://github.com/meshtag/xdsl/blob/10f1a5598320f1692d1a087b546511a3e2a991d4/tests/test_ir.py#L292)?
```
ToDo: Create this op without utilising irdl_op_definition decorator, since it tests fine grained stuff which is supposed to be used with IRDL or PDL itself.
```

Reply:
I think `without IRDL itself`makes more sense :)
Other than that, it LGTM.

Reply:
> and we need to open an issue on the xDSL repo detailing this issue/discussion.

Created [this feature request](https://github.com/xdslproject/xdsl/issues/769), perhaps we can continue this discussion there? 
(/cc @webmiche )

Reply:
I have to say, I'm not really sure that testing it without IRDL adds much here?
IRDL only generates the verifier and some accessors.
IRDL doesn't interfere at all here? Especially since we just have a `VarRegion`.
But if you feel strongly about it, feel free to change it, I don't have a strong opinion here.